services/network/networksecuritygroup/wssd.go

```diff
@@ -0,0 +1,327 @@
+
+	for _, wssdCloudNSG := range response.GetNetworkSecurityGroups() {
+		networkNSG, err := getNetworkSecurityGroup(wssdCloudNSG)
+		if err != nil {
+			return nil, err
+		}
+
+		networkdNSGs = append(networkdNSGs, *networkNSG)
+	}
+
+	return &networkdNSGs, nil
+}
+
+// getWssdNetworkSecurityGroup converts our internal representation of a networksecuritygroup (network.SecurityGroup) to the cloud network security group protobuf used by wssdcloudagent (wssdnetwork.NetworkSecurityGroup)
+func getWssdNetworkSecurityGroup(networkNSG *network.SecurityGroup, group string) (wssdCloudNSG *wssdcloudnetwork.NetworkSecurityGroup, err error) {
+
+	if len(group) == 0 {
+		return nil, errors.Wrapf(errors.InvalidGroup, "Group not specified")
+	}
+
+	if networkNSG.Name == nil {
+		return nil, errors.Wrapf(errors.InvalidConfiguration, "Missing Name for Network Security Group")
+	}
+
+	wssdCloudNSG = &wssdcloudnetwork.NetworkSecurityGroup{
+		Name:      *networkNSG.Name,
+		GroupName: group,
+	}
+
+	if networkNSG.Location != nil {
+		wssdCloudNSG.LocationName = *networkNSG.Location
+	}
+
+	if networkNSG.Tags != nil {
+		wssdCloudNSG.Tags = tags.MapToProto(networkNSG.Tags)
+	}
+
+	if networkNSG.SecurityGroupPropertiesFormat != nil {
+		nsgRules, err := getWssdNetworkSecurityGroupRules(networkNSG.SecurityRules)
+		if err != nil {
+			return nil, err
+		}
+		wssdCloudNSG.Networksecuritygrouprules = nsgRules
+	}
+
+	return wssdCloudNSG, nil
+}
+
+// getWssdNetworkSecurityGroupRules converts our internal representation of a networksecuritygroup rule (network.SecurityRule) to the cloud network security group rule protobuf used by wssdcloudagent (wssdnetwork.NetworkSecurityGroupRule)
+func getWssdNetworkSecurityGroupRules(securityRules *[]network.SecurityRule) (wssdNSGRules []*wssdcloudnetwork.NetworkSecurityGroupRule, err error) {
+	if securityRules == nil || len(*securityRules) <= 0 {
+		return
+	}
+
+	for _, rule := range *securityRules {
+		if rule.SecurityRulePropertiesFormat == nil {
+			continue
+		}
+
+		wssdCloudNSGRule := &wssdcloudnetwork.NetworkSecurityGroupRule{}
+
+		if rule.Name == nil {
+			return nil, errors.Wrapf(errors.InvalidInput, "Network Security Rule name not specified")
+		}
+		wssdCloudNSGRule.Name = *rule.Name
+
+		if rule.Description != nil {
+			wssdCloudNSGRule.Description = *rule.Description
+		}
+
+		if strings.EqualFold(string(rule.Protocol), string(network.SecurityRuleProtocolAsterisk)) {
+			wssdCloudNSGRule.Protocol = wssdcloudcommon.Protocol_All
+		} else if strings.EqualFold(string(rule.Protocol), string(network.SecurityRuleProtocolTCP)) {
+			wssdCloudNSGRule.Protocol = wssdcloudcommon.Protocol_Tcp
+		} else if strings.EqualFold(string(rule.Protocol), string(network.SecurityRuleProtocolUDP)) {
+			wssdCloudNSGRule.Protocol = wssdcloudcommon.Protocol_Udp
+		} else {
+			return nil, errors.Wrapf(errors.InvalidInput, "Unknown Protocol %s specified", rule.Protocol)
+		}
+
+		if rule.SourceAddressPrefix != nil {
```

Comment:
Since SDK exposes `SourceAddressPrefixes` we will need to handle that case as well.

Reply:
handled, along with the other plurals!